### DISCUSSION TITLE
Matching one image color to another

### DISCUSSION THREAD

**Discussion by memphis-10** [ImageMagick/ImageMagick#4367] [Digital Image Processing]
I posted another topic about specific color temp adjustments, but I believe this is separate enough to start anew:

Goal: color-balancing one image to match another.  For example: match film/sensor A to colors defined by film/sensor Z.

I've seen the thorough and invaluable chart scripts on Snibgo's page.  Amazing work.  Some of that is a little advanced for me.  I'd like to match colors between images, not necessarily to the exact chart colors.  I'm not sure if that's broken down in those scripts, but I'd need a hand deciphering if so.

Specifically: I want to extract/obtain correction values to repeat in standalone IM strings.

Initial thoughts on how to do this: 
1- shoot chart on film/sensor A (Image A)
2- shoot the same chart on film/sensor Z (Image Z)
3- color-balance Image A to to match Image Z

**This does not have to be chart-based.  I'm open to any way to make it happen.  

Thanks a ton

**Comment by fmw42** [comment]
You can do that using brightness/contrast matching using the mean and standard deviation. Both snibgo and I have scripts that do that. It matches one image to another channel by channel.

My script is bash unix. see http://www.fmwconcepts.com/imagemagick/matchimage/index.php

Snibgo's in in Windows bat script. See http://im.snibgo.com/gainbias.htm

**Reply by memphis-10** [reply]
Thank you @fmw42.  Y'all are legends.  I'm looking at your bcmatch script now.  Am I able to extract/save the values to use later?

**Reply by fmw42** [reply]
I can make it do that. But why would the linear transformation equation coefficients be of use later? Matching depends upon both images. So the match of one might not be appropriate to match some other image.

**Reply by memphis-10** [reply]
This is part of a project where I am shooting motion picture film in 35mm still cameras.  In short: the corrections will not always be image-dependent.  I'm experimenting with correcting a series of control images to obtain settings that I can apply repeatedly.

**Reply by memphis-10** [reply]
Provided there is no clipping, getting close during my IM inversion is a huge plus.  Plus doing this is part of the learning portion for my workflow.

**Comment by fmw42** [comment]
If you can make adjustments manually using Photoshop or GIMP that are global adjustments (brightness, contrast, color) rather than locally varying (noise removal, gradients, convolutions, etc), then you can use HALD (3D lookup table) images and -hald-clut to save the transformation and then use that to correct all the other images. See https://legacy.imagemagick.org/Usage/color_mods/#hald-clut.

What you do is record the operations that you do on one image in GIMP. Create a HALD image in Imagemagick saving without compression. Take that image to GIMP. Apply the same processing to it. Save again without compression. Bring the modified HALD image back to Imagemagick. Use the modified HALD image with -hald-clut to correct any other image.

**Comment by snibgo** [comment]
I suppose you have seen my [Colour checker charts](http://im.snibgo.com/colchkcha.htm) page.

>  For example: match film/sensor A to colors defined by film/sensor Z.

If you have two or more cameras, whether film or digital, then matching colours is most conveniently done by ensuring they both photograph the same colour chart. Then you can match all the shots to whichever one you consider most "correct". If you use a calibrated chart such as a Macbeth, you can match them all to an absolute standard.

I do this with digital cameras. I have no experience of matching digitised film. With digitised film, we would ideally first find the transformation from the digitised film to a linear light representation, ie where the pixel numbers are proportional to the light energy received by that part of the film. In other words, we would want to reverse the characteristic curve of the film. A standard colour chart would help that process.

If you don't have a colour chart, you would need to select frames (perhaps cropped) from the cameras that you think should match. This will be more approximate and error-prone than using charts.

Some interactive software such as darktable contains specialist modules for processing and matching colour charts.

**Comment by memphis-10** [comment]
Thank you both for taking the time to help.  I know your respective IM and image processing knowledge is massive.  I'm a decent photographer, but light years behind you in the science of correction in these terms.

@fmw42 If I use the mean deviation method, how do I save the values used?  Is that out of the question?  If so, just let me know.  In some cases, I use a HALD image as part of the processing.  Can they be stacked--meaning: can I use more than one?

@snibgo I have been studying your chart pages. Matching one chart to the other is exactly what I'm trying to do.    I'm still trying to learn how to use your scripts as BATs, where to set the variables, etc.  Once I figure out the process, is it possible to save the values for reuse in another IM string?

Thanks a ton.

**Reply by fmw42** [reply]
I could modify the output to save the transformation coefficients (a and b) for the linear transformation if that is what you want. They can be stacked, but I do not see the point, since the HALD could contain the results of the image matching if you really think it is not image dependent.

**Reply by memphis-10** [reply]
Thank you for taking the time to respond.  I'm just as into learning why- as I am how.

The reason I asked about HALD stacking is that there may be other HALDs in the string someday which I'm unable to edit.  If I use a HALD for my color-corrections, then I will need to stack.  

I may have been overly-generous with my skill in saying I can make the corrections in editing software.  I probably can, but I'm only okay at Photoshop and just now learning Gimp.  I primarily use Lightroom which has simple sliders (plus "auto") for many of the corrections.  I'm eager to explore the HALD method, but I'm still wondering if the values can be created in IM.

I see why you wonder why I would want to repeat the values.  Most immediately, I am working on color corrections for images with incorrect white balance.  Shooting a chart on a tripod under both conditions gives me an image that should yield reusable values, no?

Thanks again.

**Reply by fmw42** [reply]
You can stack HALD, I believe. I see no reason why you could not. But again they need to be global changes (brightness, contrast, color) and not involve blurring, sharpening, noise removal.

**Reply by fmw42** [reply]
You might also investigate my whitebalancing script.

**Reply by memphis-10** [reply]
Thank you.  Below Snibgo suggested I apply one HALD to another and use the third as needed.  I had the same thought this morning.  Once I create a HALD with my corrections, I will try that--as well as stacking them.

I have looked at your WB script a few times.  Whitebalancing, bcmatch, matchimage, and colorfilter are all of great interest to me for this project--and I'm very thankful you've shared these.  I'm somewhat new to script use on the Mac, side, though that's my preferred OS.  Once I traverse that learning curve, I'll have more skill in seeing how the changes take place which will set me up better for saving my own values.  Regardless of the method or results, an important part of this process is being able to find the values that work for me and reuse them.  I have some excellent starting points and will be trying your scripts along the way.

Thank you again.

**Comment by snibgo** [comment]
> Once I figure out the process, is it possible to save the values for reuse in another IM string?

Yes. On my linked page, I show how to get a 6x6 matrix or a polynomial that can be applied to any image. A saved matrix can be applied to any image with "-color-matrix". For polynomials, use "-function Polynomial".

**Reply by memphis-10** [reply]
Thank you very much.  I'm only minimally skilled at this so I'm stumbling though that page.  I'll continue learning how to use this script to modify one image to match the colors in another.  I do have a chart, but not a perfect one. It should suffice for my purposes, if not true macbeth color matching.

**Comment by snibgo** [comment]
If you have two or more Hald cluts, you can apply just the first to your image, and then the second to the result, and keep applying as many as you want.

You can get the same result by applying the second Hald to the first Hald to get another Hald, then the third Hald to the result from the first two, and keep going, and finally apply the resulting Hald to the input image.

**Reply by memphis-10** [reply]
I had this same idea when I woke up.  Amazing what some sleep can do for progress.  Thank you!
